docs/notebooks/jax-101/01-jax-basics.md

```diff
@@ -0,0 +1,375 @@
+
+A side-effect is any effect of a function that doesn't appear in its output. One example is modifying an array in place:
+
+```{code-cell} ipython3
+:id: o_YBuLQC1wPJ
+
+import numpy as np
+
+x = np.array([1, 2, 3])
+
+def in_place_modify(x):
+  x[0] = 123
+  return None
+
+in_place_modify(x)
+x
+```
+
++++ {"id": "JTtUihVZ13F6"}
+
+The side-effectful function modifies its argument, but returns a completely unrelated value. The modification is a side-effect. 
+
+This will run in Numpy. However, JAX arrays won't allow themselves to be modified in-place:
+
+```{code-cell} ipython3
+:id: u6grTYIVcZ3f
+:tags: [raises-exception]
+
+in_place_modify(jnp.array(x))  # Raises error when we cast input to jnp.ndarray
+```
+
++++ {"id": "RGqVfYSpc49s"}
+
+Helpfully, the error points us to JAX's side-effect-free way of doing the same thing via the [`jax.ops.index_*`](https://jax.readthedocs.io/en/latest/jax.ops.html#indexed-update-operators) ops. They are analogous to in-place modification by index, but create a new array with the corresponding modifications made:
+
+```{code-cell} ipython3
+:id: Rmklk6BB2xF0
+
+def jax_in_place_modify(x):
+  return jax.ops.index_update(x, 0, 123)
+
+y = jnp.array([1, 2, 3])
+jax_in_place_modify(y)
+```
+
++++ {"id": "91tn_25vdrNf"}
+
+Note that the old array was untouched, so there is no side-effect:
+
+```{code-cell} ipython3
+:id: KQGXig4Hde6T
+
+y
+```
+
++++ {"id": "d5TibzPO25qa"}
+
+Side-effect-free code is sometimes called *functionally pure*, or just *pure*.
+
+Isn't the pure version less efficient? Strictly, yes; we are creating a new array. However, as we will explain in the next guide, JAX computations are often compiled before being run using another program transformation, `jax.jit`. If we don't use the old array after modifying it 'in place' using `jax.ops.index_update()`, the compiler can recognise that it can in fact compile to an in-place modify, resulting in efficient code in the end.
+
+Of course, it's possible to mix side-effectful Python code and functionally pure JAX code, and we will touch on this more later. As you get more familiar with JAX, you will learn how and when this can work. As a rule of thumb, however, any functions intended to be transformed by JAX should avoid side-effects, and the JAX primitives themselves will try to help you do that.
+
+We will explain other places where the JAX idiosyncracies become relevant as they come up. There is even a colab that focuses entirely on getting used to the functional programming style of handling state: [Part 7: Problem of State](https://colab.research.google.com/github/google/jax/blob/master/docs/notebooks/jax-101/07-state.ipynb). However, if you're impatient, you can find a [summary of JAX's sharp edges](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) in the JAX docs.
```

Comment:
```suggestion
We will explain other places where the JAX idiosyncracies become relevant as they come up. There is even a Colab that focuses entirely on getting used to the functional programming style of handling state: [Part 7: Problem of State](https://colab.research.google.com/github/google/jax/blob/master/docs/notebooks/jax-101/07-state.ipynb). However, if you're impatient, you can find a [summary of JAX's sharp edges](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) in the JAX docs.
```

Reply:
(I think the 🔪 sharp edges/common gotchas 🔪 notebook should be right below the Quickstart in terms of importance 😄 in TOC)